### PR TITLE
Improvements of "mxpy testnet". Re-brand to "localnet".

### PR DESCRIPTION
**Pull request by andreibancioiu**
:exclamation:  **Breaking change:** **new configuration schema** (see below). Flow slightly altered. New commands added.

 - Refactoring, partial re-design.
 - Allow one to use pre-built binaries of `node`, `seednode` and `proxy`.
 - Allow one to configure round duration, epoch duration etc.

## Re-branding

`mxpy testnet` :arrow_right: `mxpy localnet`
`testnet.toml` :arrow_right: `localnet.toml`

## Commands

:green_circle: New command, creates, builds, configures a localnet (without starting it, though) - this calls `new` ,`prerequisites`, `build` and `config` in one go.

```
mxpy localnet setup
```

:green_circle: New command: creates a `localnet.toml` configuration file.

```
mxpy localnet new
```

:green_circle:  The clean command stays the same:

```
mxpy localnet clean
```

:green_circle: Fetch repositories and other dependencies.

```
mxpy localnet prerequisites
```

:green_circle: New command: will build the software.

```
mxpy localnet build
```

:green_circle: NO BUILD, just copying of files & folders & patching the config files

```
mxpy localnet config
```

:green_circle: Start the network (can be stopped at any time by killing the `mxpy` process).

```
mxpy localnet start
```

## :gear:  New configuration schema. 

Below is an example of **`localnet.toml`**. Software `resolution` can be `remote` or `local`.

```
[general]
log_level = "*:DEBUG"
genesis_delay_seconds = 10
rounds_per_epoch = 100
round_duration_milliseconds = 6000

[metashard]
consensus_size = 1
...

[shards]
num_shards = 1
...

[networking]
...
port_proxy = 7950
...

[software.mx_chain_go]
resolution = "remote"
archive_url = "https://github.com/multiversx/mx-chain-go/archive/refs/heads/master.zip"
archive_download_folder = "~/multiversx-sdk/localnet_software_remote/downloaded/mx-chain-go"
archive_extraction_folder = "~/multiversx-sdk/localnet_software_remote/extracted/mx-chain-go"
local_path = "~/multiversx-sdk/localnet_software_local/mx-chain-go"

[software.mx_chain_proxy_go]
resolution = "remote"
archive_url = "https://github.com/multiversx/mx-chain-proxy-go/archive/refs/heads/master.zip"
archive_download_folder = "~/multiversx-sdk/localnet_software_remote/downloaded/mx-chain-proxy-go"
archive_extraction_folder = "~/multiversx-sdk/localnet_software_remote/extracted/mx-chain-proxy-go"
local_path = "~/multiversx-sdk/localnet_software_local/mx-chain-proxy-go"
```

### CODE REVIEW

File: multiversx_sdk_cli/localnet/config_software.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        self.mx_chain_proxy_go.override(other.get("mx_chain_proxy_go", {}))
+
+    def to_dictionary(self) -> Dict[str, Any]:
+        return {
+            "mx_chain_go": self.mx_chain_go.to_dictionary(),
+            "mx_chain_proxy_go": self.mx_chain_proxy_go.to_dictionary(),
+        }
+
+
+class SoftwarePiece(ConfigPart):
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply multiversx_sdk_cli/localnet/config_software.py:36]: Renamed.

File: multiversx_sdk_cli/localnet/genesis_smart_contracts_json.py [hunk 2]
```diff
@@ -1,16 +1,16 @@
...
-def patch(data: Any, testnet_config: TestnetConfiguration):
+def patch(data: Any, config: ConfigRoot):
     owner = genesis.get_owner_of_genesis_contracts()
 
     delegation_config = data[0]
     dns_config = data[1]
 
     delegation_config["owner"] = owner.address.bech32()
     dns_config["owner"] = owner.address.bech32()
-    # registration price = 100 atoms of eGLD
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply multiversx_sdk_cli/localnet/genesis_smart_contracts_json.py:15]: Fixed.

**Review by camilbancioiu — APPROVED** [review]

File: multiversx_sdk_cli/localnet/libraries.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+
+def copy_libraries(source: Path, destination: Path):
+    libraries: List[Path] = list(source.glob("*.dylib")) + list(source.glob("*.so"))
+
+    for library in libraries:
+        logger.debug(f"Copying {library} to {destination}")
+
+        shutil.copy(library, destination)
+        os.chmod(destination / library.name, 0o755)
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply multiversx_sdk_cli/localnet/libraries.py:17]: Fixed.

File: multiversx_sdk_cli/cli_localnet.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    help_config_file = "An optional configuration file describing the localnet"
+    parser.add_argument("--configfile", type=Path, required=False, default=Path("localnet.toml"), help=help_config_file)
+
+
+def localnet_new(args: Any):
+    logger.info("New localnet (creating configuration file)...")
+
+    step_new.new_config(args.configfile)
+
+    logger.info("New localnet configuration file created (or already existing). Make sure to inspect it. Now, you can run 'mxpy localnet prerequisites' to fetch localnet prerequisites.")
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply multiversx_sdk_cli/cli_localnet.py:106]: Sure! Added, using https://rich.readthedocs.io/.

File: multiversx_sdk_cli/localnet/config_networking.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+                 port_seednode: int,
+                 p2p_id_seednode: str,
+                 port_proxy: int,
+                 port_first_observer: int,
+                 port_first_observer_rest_api: int,
+                 port_first_validator: int,
+                 port_first_validator_rest_api: int):
+        self.host: str = host
+        self.port_seednode: int = port_seednode
+        self.p2p_id_seednode: str = p2p_id_seednode
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply multiversx_sdk_cli/localnet/config_networking.py:18]: Good catch! Fixed.

**Review by bogdan-rosianu — APPROVED** [review]

File: multiversx_sdk_cli/localnet/constants.py [hunk 6]
```diff
@@ -1,2 +1,6 @@
+import stat
+
 METACHAIN_ID = 4294967295
 NETWORK_MONITORING_INTERVAL_IN_SECONDS = 1
+# Read, write and execute by owner, read and execute by group and others
+FILE_MODE_EXECUTABLE = stat.S_IRUSR | stat.S_IWUSR | stat.S_IXUSR | stat.S_IRGRP | stat.S_IXGRP | stat.S_IROTH | stat.S_IXOTH
```
Comment (camilbancioiu) [inline multiversx_sdk_cli/localnet/constants.py:6]: :+1:

**Review by camilbancioiu — APPROVED** [review]

**Pull request merged**